source/common/upstream/load_balancer_impl.cc

```diff
@@ -426,55 +423,43 @@ RoundRobinLoadBalancer::RoundRobinLoadBalancer(
       [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
 }
 
-void RoundRobinLoadBalancer::refresh(uint32_t priority) {
+void EdfLoadBalancerBase::initialize() {
+  for (uint32_t priority = 0; priority < priority_set_.hostSetsPerPriority().size(); ++priority) {
+    refresh(priority);
+  }
+}
+
+void EdfLoadBalancerBase::refresh(uint32_t priority) {
   const auto add_hosts_source = [this](HostsSource source, const HostVector& hosts) {
-    bool weighted = false;
+    // Nuke existing scheduler if it exists.
+    auto& scheduler = scheduler_[source] = Scheduler{};
+    refreshHostSource(source);
 
+    // Populate scheduler with host list.
+    // TODO(mattklein123): We must build the EDF schedule even if all of the hosts are currently
+    // weighted 1. This is because currently we don't refresh host sets if only weights change.
+    // We should probably change this to refresh at all times. See the comment in
```

Comment:
The cost here is the rebuild overhead. If we used a non-STL priority queue data structure, we could possibly just rebuild with the modified hosts at a lower cost.

Reply:
Also memory, but yeah.